### PR TITLE
[Android Auto] Use the RouteLineComponent and RouteArrowComponent

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Here's an alternative for https://github.com/mapbox/mapbox-navigation-android/pull/5910

### CODE REVIEW

**Comment by Guardiola31337** [comment]
@Zayankovsky could you take a look at this when you have a chance? 🙏

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt [hunk 1]
```diff
@@ -34,37 +34,38 @@ import com.mapbox.maps.MapboxExperimental
...
 @MapboxExperimental
+@ExperimentalPreviewMapboxNavigationAPI
 class CarRoutePreviewScreen(
     private val routePreviewCarContext: RoutePreviewCarContext,
     private val placeRecord: PlaceRecord,
     private val navigationRoutes: List<NavigationRoute>,
     private val placesLayerUtil: PlacesListOnMapLayerUtil = PlacesListOnMapLayerUtil(),
 ) : Screen(routePreviewCarContext.carContext) {
 
-    private val routesProvider = PreviewRoutesProvider(navigationRoutes)
```
Comment (Zayankovsky) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: Why did you remove this? We don't want to set routes during route preview (too many issues with this approach), thus we need to store the routes locally.

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: The `RouteLineComponent` will only use routes set by `MapboxNavigation`. There is also an issue with not setting the route during route preview, route progress does not work. So if you try navigating during route preview it will go stale.

But yeah I'll look to update this. RouteLineComponent will need to be able to accept the route outside of MapboxNavigation

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: I'm making this change blocked by https://github.com/mapbox/mapbox-navigation-android/issues/5990

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: > too many issues with this approach

Is there a list of issues somewhere? I don't know how you could make route preview a state of active navigation otherwise

Comment (abhishek1508) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: > I don't know how you could make route preview a state of active navigation otherwise

Why do we want to do this? Is there an actual customer request to do so?

Comment (abhishek1508) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: I would not worry about route progress not working in route preview state because that is something that should not happen. Route preview allows me to decide the route I want to choose before actually starting the navigation. In that case, why would I want `routeProgress` in Route Preview mode.

Comment (abhishek1508) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: What's the reason for implementing >I'm making this change blocked by https://github.com/mapbox/mapbox-navigation-android/issues/5990

in this PR? Can we keep the implementation separate so that we don't make the PR complicated and keep the solution to what the original ticket is asking for?

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRoutePreviewScreen.kt:51]: Android Auto cannot use RouteLineComponent for the same reason that 1tap cannot use RouteLineComponent, as Zayankovsky is warning by the first comment in the thread.

RoutePreview cannot set the MapboxNavigation route. But the RouteLineComponent is essentially an `ActiveGuiidanveRouteLineComponent`

**Comment by abhishek1508** [comment]
Blocked on https://github.com/mapbox/mapbox-navigation-android/issues/5990

**Comment by abhishek1508** [comment]
#5990 is fixed and this PR should now be unblocked @kmadsen

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt [hunk 2]
```diff
@@ -1,165 +1,95 @@
...
-                        routeLineView.hideAlternativeRoutes(style)
-                    }
-                }
+    private fun attachRouteArrow(mapboxCarMapSurface: MapboxCarMapSurface) {
+        routeArrowComponent = RouteArrowComponent(
+            mapboxMap = mapboxCarMapSurface.mapSurface.getMapboxMap(),
+            routeArrowOptions = routeArrowOptions
+                ?: RouteArrowOptions.Builder(mapboxCarMapSurface.carContext)
+                    .build(),
+        ).also { routeLineComponent ->
```
Comment (abhishek1508) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt:69]:
```suggestion
        ).also { routeArrowComponent ->
```

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt [hunk 3]
```diff
@@ -1,165 +1,95 @@
...
+            mapboxCarMapSurface.getStyleAsync {
+                MapboxNavigationApp.registerObserver(routeLineComponent)
```
Comment (abhishek1508) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt:71]:
```suggestion
                MapboxNavigationApp.registerObserver(routeArrowComponent)
```

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt [hunk 4]
```diff
@@ -164,7 +146,7 @@ class CarNavigationCamera internal constructor(
         )
 
         mapboxNavigation.registerLocationObserver(locationObserver)
-        routesProvider.registerRoutesListener(routesListener)
+        mapboxNavigation.registerRoutesObserver(routesObserver)
```
Comment (Zayankovsky) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:149]: I don't think observing routes from `MapboxNavigation` is correct here, since routes are not set there during route preview. I believe that's why I can't see the whole route in the preview state with these changes.

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt [hunk 5]
```diff
@@ -1,165 +1,95 @@
...
-
-            scope.coroutineContext.cancelChildren()
-            scope.launch {
-                mainCarContext.routeAlternativesEnabled.collect { enabled ->
-                    if (enabled) {
-                        routeLineView.showAlternativeRoutes(style)
-                    } else {
-                        routeLineView.hideAlternativeRoutes(style)
-                    }
-                }
```
Comment (Zayankovsky) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt:136]: Where is this handled now? Looks like we don't have that logic anymore.

**Review by Zayankovsky — CHANGES_REQUESTED** [review]

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt (hunk 4 above)
Comment (abhishek1508) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:149]: This is needed for the camera to function in active guidance state. However, I agree that camera is missing observing routes from route preview state. This is how it's been done in the Drop-In UI
```
coroutineScope.launch {
            val routesFlow = mapboxNavigation.flowRoutesUpdated()
                .map { it.navigationRoutes }
                .stateIn(
                    coroutineScope,
                    SharingStarted.WhileSubscribed(),
                    mapboxNavigation.getNavigationRoutes()
                )

            val routePreview = store.select { it.previewRoutes }
            val navigation = store.select { it.navigation }
            combine(
                routePreview,
                routesFlow,
                navigation
            ) { previewRoutes, navigationRoutes, navigationState ->
                val routes = if (previewRoutes is RoutePreviewState.Ready) {
                    previewRoutes.routes
                } else if (navigationRoutes.isNotEmpty()) {
                    navigationRoutes
                } else {
                    emptyList()
                }
                when (routes.isNotEmpty()) {
                    true -> {
                        viewportDataSource.onRouteChanged(routes.first())
                        viewportDataSource.evaluate()
                        when (navigationState) {
                            NavigationState.ActiveNavigation,
                            NavigationState.Arrival -> {
                                store.dispatch(SetCameraMode(TargetCameraMode.Following))
                            }
                            else -> {
                                store.dispatch(SetCameraMode(TargetCameraMode.Overview))
                            }
                        }
                    }
                    false -> {
                        viewportDataSource.clearRouteData()
                        viewportDataSource.evaluate()
                    }
                }
            }.collect()
        }
```
It always looks at the route from both routePreview and onRoutesChanged. Whenever a route is available, the camera viewport updates.

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:149]: Yeah you're right, camera needs to share a common route provider. 

The more generic `RoutesProvider` feels better because of this 🤔

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt (hunk 5 above)
Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt:136]: Good catch!

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/ActiveRouteLineComponentContract.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package com.mapbox.navigation.ui.maps.internal.ui
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.MapboxNavigation
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOf
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class ActiveRouteLineComponentContract : RouteLineComponentContract {
```
Comment (kmadsen) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/ActiveRouteLineComponentContract.kt:10]: Exposing this so we can use it downstream (internal for android auto)

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt (hunk 5 above)
Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/preview/CarRouteLine.kt]: https://github.com/mapbox/mapbox-navigation-android/pull/5912/files#diff-6198cd76f0edf02371f3b1f7d936ab739daddc8f9063c26b62459bd42f8452a2R81

Ok added it back in

File: libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt [hunk 7]
```diff
@@ -16,35 +14,43 @@ import com.mapbox.maps.extension.androidauto.DefaultMapboxCarMapGestureHandler
...
+class CarNavigationCamera(
     private val mapboxNavigation: MapboxNavigation,
     private val initialCarCameraMode: CarCameraMode,
     private val alternativeCarCameraMode: CarCameraMode?,
-    private val routesProvider: RoutesProvider,
     private val initialCameraOptions: CameraOptions? = CameraOptions.Builder()
         .zoom(DEFAULT_INITIAL_ZOOM)
-        .build()
+        .build(),
+    private val contract: RouteLineComponentContract = MapboxRouteLineComponentContract(),
```
Comment (kmadsen) [inline libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:47]: This does not feel right.. But it's a simple way to share the route with the route line

Right now the `CameraComponent` from dropin-ui does not use a contract with shared route preview state.

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:47]: Am considering, bringing back the `RoutesProvider` interface because this contract does not make sense

Comment (abhishek1508) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:47]: `RouteLineContract` should not be used in `Camera`. Unless `CameraComponent` in Drop-In is moved outside, AA should continue to use `RoutesProvider` for their own camera implementation.

Comment (kmadsen) [reply libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/navigation/CarNavigationCamera.kt:47]: RoutesProvider is not making it simpler, adds an extra layer of glue

**Comment by Zayankovsky** [comment]
Overview camera doesn't work in Active Guidance state:

https://user-images.githubusercontent.com/2395284/182704945-6213481b-4199-47d6-8e49-25b85ae62f7b.mov

**Comment by kmadsen** [comment]
Not taking this approach
